[Merged by Bors] - feat(linear_algebra): determinant of vectors in a basis

PatrickMassot opened the pull request:
From the sphere eversion project, define the determinant of a family of vectors with respects to a basis. 

The main result is `is_basis.iff_det` asserting a family of vectors is a basis iff its determinant in some basis is invertible.

Also renames `equiv_fun_basis` to `is_basis.equiv_fun` and `equiv_fun_basis_symm_apply` to `is_basis.equiv_fun_symm_apply`, in order to use dot notation.

Co-authored-by: Anne Baanen t.baanen@vu.nl

---
<!-- put comments you want to keep out of the PR commit here -->

Note that this file `linear_algebra.matrix` is becoming a mess. We should probably move a few things and add headers, but I didn't want to do this and add content at the same time (and also I'm too ~~lazy~~ busy right now).

PatrickMassot committed:
feat(linear_algebra): determinant of vectors in a basis

Vierkantor commented on src/linear_algebra/matrix.lean:
```diff
@@ -294,8 +313,49 @@ end
+/-- From a basis `e : ι → M` and a family of vectors `v : ι → M`, make the matrix whose columns
+are the vectors `v i` written in the basis `e`. -/
+def is_basis.to_matrix {e : ι → M} (he : is_basis R e) (v : ι → M) : matrix ι ι R :=
+  linear_equiv_matrix he he (he.constr v)
```
```suggestion
linear_equiv_matrix he he (he.constr v)
```

Vierkantor commented on src/linear_algebra/matrix.lean:
```diff
@@ -335,8 +395,43 @@ def linear_equiv.of_is_unit_det {f : M →ₗ[R] M'} {hv : is_basis R v} {hv' :
+
+lemma is_basis.det_apply (v : ι → M) : he.det v = det (he.to_matrix v) := rfl
+
+lemma is_bases.det_self : he.det e = 1 :=
```
```suggestion
lemma is_basis.det_self : he.det e = 1 :=
```

Vierkantor commented on src/linear_algebra/matrix.lean:
```diff
@@ -335,8 +395,43 @@ def linear_equiv.of_is_unit_det {f : M →ₗ[R] M'} {hv : is_basis R v} {hv' :
+lemma is_bases.det_self : he.det e = 1 :=
+by simp [he.det_apply]
+
+lemma is_basis.iff_det (v : ι → M) : is_basis R v ↔ is_unit (he.det v) :=
```
`v` should be inferrable here, right?
```suggestion
lemma is_basis.iff_det {v : ι → M} : is_basis R v ↔ is_unit (he.det v) :=
```

Vierkantor commented on src/linear_algebra/matrix.lean:
```diff
@@ -294,8 +313,49 @@ end
+
+open function matrix
+
+/-- From a basis `e : ι → M` and a family of vectors `v : ι → M`, make the matrix whose columns
+are the vectors `v i` written in the basis `e`. -/
+def is_basis.to_matrix {e : ι → M} (he : is_basis R e) (v : ι → M) : matrix ι ι R :=
```
Would it be hard to make this a linear map (or even a linear equivalence)?

PatrickMassot replied:
I did it. This required a bit of refactoring and some API holes filling.

Vierkantor commented on src/linear_algebra/matrix.lean:
```diff
@@ -15,16 +16,34 @@ to matrices. This defines a linear equivalence between linear maps
 between finite-dimensional vector spaces and matrices indexed by
 the respective bases.
 
+It also defines the trace of a endomorphism, and the determinant of a family of vectors with
```
```suggestion
It also defines the trace of an endomorphism, and the determinant of a family of vectors with
```

Vierkantor commented on src/linear_algebra/matrix.lean:
```diff
@@ -15,16 +16,34 @@ to matrices. This defines a linear equivalence between linear maps
+* `alg_equiv_matrix`: given a basis indexed by `n`, the `R`-algebra equivalence between
+  `R`-endomorphisms of `M` and `matrix n n R`
+* `matrix.trace`: the trace of a square matrix
+* `linear_map.trace`: the trace of a endomorphism
```
```suggestion
* `linear_map.trace`: the trace of an endomorphism
```

Vierkantor reviewed (changes requested):
Thanks :tada:

PatrickMassot committed:
Apply suggestions from code review

Co-authored-by: Vierkantor <Vierkantor@users.noreply.github.com>

PatrickMassot committed:
Make `to_matrix` and equiv

Also renames `equiv_fun_basis` to `is_basis.equiv_fun`
in order to use dot notation.

Vierkantor reviewed (approved):
Thanks for the work! Since I co-authored some of the definitions, I'll invite someone else to take a look as well.

bryangingechen commented on src/linear_algebra/basis.lean:
```diff
@@ -976,7 +980,7 @@ variables [fintype ι] (h : is_basis R v)
 /-- A module over `R` with a finite basis is linearly equivalent to functions from its basis to `R`.
 -/
-def equiv_fun_basis  : M ≃ₗ[R] (ι → R) :=
+def is_basis.equiv_fun  : M ≃ₗ[R] (ι → R) :=
```
```suggestion
def is_basis.equiv_fun : M ≃ₗ[R] (ι → R) :=
```

robertylewis committed:
Update src/linear_algebra/basis.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

robertylewis committed:
fix build

bryangingechen commented on src/analysis/normed_space/finite_dimension.lean:
```diff
@@ -74,13 +74,13 @@ all norms are equivalent in finite dimension.
 continuous, in `linear_map.continuous_of_finite_dimensional`. -/
 lemma continuous_equiv_fun_basis {ι : Type v} [fintype ι] (ξ : ι → E) (hξ : is_basis 𝕜 ξ) :
-  continuous (equiv_fun_basis hξ) :=
+  continuous (is_basis.equiv_fun hξ) :=
```
This should work too, right?
```suggestion
  continuous hξ.equiv_fun :=
```

bryangingechen commented on src/linear_algebra/basis.lean:
```diff
@@ -985,17 +989,17 @@ linear_equiv.trans (module_equiv_finsupp h)
 /-- A module over a finite ring that admits a finite basis is finite. -/
 def module.fintype_of_fintype [fintype R] : fintype M :=
-fintype.of_equiv _ (equiv_fun_basis h).to_equiv.symm
+fintype.of_equiv _ (h.equiv_fun).to_equiv.symm
```
```suggestion
fintype.of_equiv _ h.equiv_fun.to_equiv.symm
```

bryangingechen commented on src/linear_algebra/basis.lean:
```diff
@@ -985,17 +989,17 @@ linear_equiv.trans (module_equiv_finsupp h)
 theorem module.card_fintype [fintype R] [fintype M] :
   card M = (card R) ^ (card ι) :=
-calc card M = card (ι → R)    : card_congr (equiv_fun_basis h).to_equiv
+calc card M = card (ι → R)    : card_congr (h.equiv_fun).to_equiv
```
```suggestion
calc card M = card (ι → R)    : card_congr h.equiv_fun.to_equiv
```

PatrickMassot committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

bryangingechen commented on src/algebra/big_operators/pi.lean:
```diff
@@ -37,6 +37,11 @@ lemma finset.prod_apply {α : Type*} {β : α → Type*} {γ} [∀a, comm_monoid
   by rw [pi.multiset_prod_apply, multiset.map_map]
 
+@[simp, to_additive]
+lemma fintype.prod_apply  {α : Type*} {β : α → Type*} {γ : Type*} [fintype γ]
```
```suggestion
lemma fintype.prod_apply {α : Type*} {β : α → Type*} {γ : Type*} [fintype γ]
```

bryangingechen commented on src/linear_algebra/matrix.lean:
```diff
@@ -15,16 +16,34 @@ to matrices. This defines a linear equivalence between linear maps
+* `alg_equiv_matrix`: given a basis indexed by `n`, the `R`-algebra equivalence between
+  `R`-endomorphisms of `M` and `matrix n n R`
+* `matrix.trace`: the trace of a square matrix
+* `linear_map.trace`: the trace of an endomorphism
```
```suggestion
* `linear_map.trace`: the trace of an endomorphism
* `is_basis.to_matrix`: the matrix whose columns are a given family of vectors in a given basis
* `is_basis.to_matrix_equiv`: given a basis, the linear equivalence between families of vectors
  and matrices arising from `is_basis.to_matrix`
```

bryangingechen reviewed (approved):
I'm not too familiar with the linear algebra part of the library, but this looks good to me. 

I do agree that `linear_algebra/matrix.lean` does need to be cleaned up or split up into multiple files. The file could also use sectioning module doc strings.

jcommelin commented on src/linear_algebra/basic.lean:
```diff
@@ -1709,6 +1709,9 @@ rfl
 @[simp] theorem map_zero : e 0 = 0 := e.to_linear_map.map_zero
 @[simp] theorem map_smul (c : R) (x : M) : e (c • x) = c • e x := e.map_smul' c x
 
+@[simp] lemma map_sum [fintype ι] (u : ι → M) : e (∑ i, u i) = ∑ i, e (u i) :=
```
Shouldn't this rather be done more generally for sums over finsets?

PatrickMassot committed:
generalize lemma following Johan's suggestion

PatrickMassot committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>
